week5_policy_based/practice_reinforce_pytorch.ipynb

```diff
@@ -0,0 +1,442 @@
+    "\n",
+    "    # predict logits, probas and log-probas using an agent.\n",
+    "    logits = model(states)\n",
+    "    probs = nn.functional.softmax(logits, -1)\n",
+    "    log_probs = nn.functional.log_softmax(logits, -1)\n",
+    "\n",
+    "    assert all(isinstance(v, torch.Tensor) for v in [logits, probs, log_probs]), \\\n",
+    "        \"please use compute using torch tensors and don't use predict_probs function\"\n",
+    "\n",
+    "    # select log-probabilities for chosen actions, log pi(a_i|s_i)\n",
+    "    log_probs_for_actions = torch.sum(\n",
+    "        log_probs * to_one_hot(actions, env.action_space.n), dim=1)\n",
+    "   \n",
+    "    # Compute loss here. Don't forgen entropy regularization with `entropy_coef` \n",
+    "    entropy = <YOUR CODE>\n",
+    "    loss = <YOUR CODE>\n",
+    "\n",
+    "    # Gradient descent step\n",
+    "    <YOUR CODE>\n",
+    "\n",
+    "    # technical: return session rewards to print them later\n",
+    "    return np.sum(rewards)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### The actual training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for i in range(100):\n",
+    "    rewards = [train_on_session(*generate_session(env))\n",
+    "               for _ in range(100)]  # generate new sessions\n",
+    "    print(\"mean reward:%.3f\" % (np.mean(rewards)))\n",
+    "    if np.mean(rewards) > 300:\n",
+    "        print(\"You Win!\")  # but you can train even further\n",
+    "        break"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Video"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# record sessions\n",
+    "import gym.wrappers\n",
+    "monitor_env = gym.wrappers.Monitor(gym.make(\"CartPole-v0\"),\n",
+    "                           directory=\"videos\", force=True)\n",
+    "sessions = [generate_session(monitor_env) for _ in range(100)]\n",
+    "monitor_env.close()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# show video\n",
+    "from IPython.display import HTML\n",
+    "import os\n",
+    "\n",
+    "video_names = list(\n",
+    "    filter(lambda s: s.endswith(\".mp4\"), os.listdir(\"./videos/\")))\n",
+    "\n",
+    "HTML(\"\"\"\n",
+    "<video width=\"640\" height=\"480\" controls>\n",
+    "  <source src=\"{}\" type=\"video/mp4\">\n",
+    "</video>\n",
+    "\"\"\".format(\"./videos/\" + video_names[-1]))  # this may or may not be the _last_ video. Try other indices"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from submit import submit_cartpole_pytorch\n",
```

Comment:
It's not critical, but it would be nice to unify submission functions for PyTorch and TF.

Reply:
Sure